File: 1-working-with-list/p17_split.ml

let split list count =
  let rec aux (a1, a2) curr = function
    | [] -> (List.rev a1, List.rev a2)
    | h :: t ->
      if curr = 0 then
        aux (a1, h::a2) curr t
      else
        aux (h::a1, a2) (curr-1) t
  in aux ([], []) count list;;

let split list count =
  let rec aux (a1, a2) curr = function
    | [] -> (List.rev a1, [])
    | h :: t ->
      if curr = 0 then
        (List.rev a1, t)
      else
        aux (h::a1, a2) (curr-1) t
  in aux ([], []) count list;;

split ["a";"b";"c";"d";"e";"f";"g";"h";"i";"j"] 3;;
split ["a";"b";"c";"d"] 5;;
split ["a";"b";"c";"d"; "e"] 5;;

File: 1-working-with-list/p17_split.scala
import scala.annotation.tailrec

object Main {

  def main(args: Array[String]) {
    val l = List("a", "b", "c", "d", "e", "f", "g", "h", "i", "j")
    println(split(l, 3))
    println(split(l, 0))
    println(split(l, -1))
    val m = List("a", "b", "c", "d")
    println(split(m, 5))
  }

  def split[T](list: List[T], n: Int): List[List[T]] = {
    @tailrec
    def aux(l: List[T], c: Int, acc: List[T]): List[List[T]] = {
      l match {
        case Nil => List(acc.reverse, List())
        case hd :: tail =>
          if(c <= 0)
            List(acc.reverse, l)
          else
            aux(tail, c-1, hd :: acc)
      }
    }
    aux(list, n, List())
  }
}
